tests/fuser_mqubit_test.cc

```diff
@@ -0,0 +1,295 @@
+                  std::uniform_real_distribution<double>& distr,
+                  std::mt19937& rgen, unsigned& n,
+                  std::vector<unsigned>& available,
+                  std::vector<DummyGate>& circuit) {
+  double r = distr(rgen);
+
+  if (r < p1) {
+    circuit.push_back(CreateDummyGate(time, GenQubits(1, rgen, n, available)));
+  } else if (r < p2) {
+    circuit.push_back(CreateDummyGate(time, GenQubits(2, rgen, n, available)));
+  } else if (r < p3) {
+    circuit.push_back(CreateDummyGate(time, GenQubits(3, rgen, n, available)));
+  } else if (r < p4) {
+    circuit.push_back(CreateDummyGate(time, GenQubits(4, rgen, n, available)));
+  } else if (r < p5) {
+    circuit.push_back(CreateDummyGate(time, GenQubits(5, rgen, n, available)));
+  } else if (r < p6) {
+    circuit.push_back(CreateDummyGate(time, GenQubits(6, rgen, n, available)));
+  } else if (r < pc) {
+    auto qs = GenQubits(1 + rgen() % 3, rgen, n, available);
+    auto cqs = GenQubits(1 + rgen() % 3, rgen, n, available);
+    circuit.push_back(
+        CreateDummyControlledGate(time, std::move(qs), std::move(cqs)));
+  } else if (r < pm) {
+    unsigned num_mea_gates = 0;
+    unsigned max_num_mea_gates = 1 + rgen() % 5;
+
+    while (n > 0 && num_mea_gates < max_num_mea_gates) {
+      unsigned k = 1 + rgen() % 12;
+      if (k > n) k = n;
+      circuit.push_back(
+          CreateDummyMeasurementGate(time, GenQubits(k, rgen, n, available)));
+      ++num_mea_gates;
+    }
+  }
+
+  if (r < p6 && distr(rgen) < pu) {
+    circuit.back().unfusible = true;
+  }
+}
+
+std::vector<DummyGate> GenerateRandomCircuit1(unsigned num_qubits,
+                                              unsigned depth) {
+  std::vector<DummyGate> circuit;
+  circuit.reserve(depth);
+
+  std::mt19937 rgen(1);
+  std::uniform_real_distribution<double> distr(0, 1);
+
+  std::vector<unsigned> available(num_qubits, 0);
+
+  for (unsigned time = 0; time < depth; ++time) {
+    for (unsigned k = 0; k < num_qubits; ++k) {
+      available[k] = k;
+    }
+
+    unsigned n = num_qubits;
+
+    AddToCircuit(time, distr, rgen, n, available, circuit);
+  }
+
+  return circuit;
+}
+
+std::vector<DummyGate> GenerateRandomCircuit2(unsigned num_qubits,
+                                              unsigned depth,
+                                              unsigned max_fused_gate_size) {
+  std::vector<DummyGate> circuit;
+  circuit.reserve(num_qubits * depth);
+
+  std::mt19937 rgen(1);
+  std::uniform_real_distribution<double> distr(0, 1);
+
+  std::vector<unsigned> available(num_qubits, 0);
+
+  for (unsigned time = 0; time < depth; ++time) {
+    for (unsigned k = 0; k < num_qubits; ++k) {
+      available[k] = k;
+    }
+
+    unsigned n = num_qubits;
+
+    while (n > max_fused_gate_size) {
+      AddToCircuit(time, distr, rgen, n, available, circuit);
+    }
+  }
+
+  return circuit;
+}
+
+template <typename FusedGate>
+bool TestFusedGates(unsigned num_qubits,
```

Comment:
If I understand it right, this will verify that the circuit in question is valid (no timing conflicts, measurements are isolated, and the input gates are only in the fused result once).

Could you add a test verifying that a circuit with 3+ qubit fusing can be run (e.g. with `run_qsim`) and that the results match those of a circuit run with basic fusion?

Reply:
Yes, that is right. I added a test that uses `ApplyGate` (no gate fusion) and `ApplyFusedGate` (gate fusion). I did not use basic fusion as it might get deprecated at some point.